Tweek.JPad.Tests/Matcher.Tests.fs

```diff
@@ -285,4 +282,92 @@ type ``Matcher tests`` ()=
+        let validate = validator """{"Countries": {"$all": { "countryCode":1 } }}"""
+        let validateNotExists = validator """{"Countries": {"$all": { "name": "Canada" } }}"""
+        let validateOp = validator """{"Countries": {"$all": { "countryCode": { "$lt":2, "$gt":0 } } }}"""
+        let validateOpNotExists = validator """{"Countries": {"$all": { "countryCode": { "$lt":1, "$gt":0 } } }}"""
+        let validatePrimitive = validator """{"Countries": { "$all": 1 }}"""
+        validate (context [("Countries", countries);])  |> should equal true
+        validateNotExists (context [("Countries", countries);])  |> should equal false
+        validateOp (context [("Countries", countries);])  |> should equal true
+        validateOpNotExists (context [("Countries", countries);])  |> should equal false
+        validatePrimitive (context [("Countries", countries);])  |> should equal false
+        validate (context [("Countries", JsonValue.Array([||]));])  |> should equal true
+        validate (context [("Countries", JsonValue.Null);])  |> should equal false
+        validate (context [("Countries", JsonValue.String(""));])  |> should equal false
+        validate (context [("Countries", JsonValue.Boolean(false));])  |> should equal false
+
+    [<Fact>]
+    member test.``Array comparers - all with comparers``() =
+        let devices = JsonValue.Parse """[{"name":"iPhone","AgentVersion":"0.1.1"},{"name":"Samsung","AgentVersion":"0.1.2"}]"""
+        let comparers = dict([("version", new ComparerDelegate(fun x -> Version.Parse(x) :> IComparable))])
+        let validate = validatorWithComparers """{"Devices": {"all": {"AgentVersion": {"$compare": "version", "gt": "0.1.0" }} }}""" comparers
+        let validateNotExist = validatorWithComparers """{"Devices": {"all": {"AgentVersion": {"$compare": "version", "$gt": "0.3.0" }} }}""" comparers   
+        validate (context [("Devices", devices);])  |> should equal true
+        validateNotExist (context [("Devices", devices);])  |> should equal false
+        
+    [<Fact>]
+    member test.``Array comparers - all primitives``() =
+        let ages = JsonValue.Parse """[35,35,35]"""
+        let validate = validator """{"Ages": {"$all": 35}}"""
+        let validateNotExist = validator """{"Ages": {"$all": 45}}"""
+        let validateOp = validator """{"Ages": {"$all": { "$gt": 10 }}}"""
+        let validateOpNotExist = validator """{"Ages": {"$all": { "$lt": 10 }}}"""
+        validate (context [("Ages", ages);])  |> should equal true
+        validateNotExist (context [("Ages", ages);])  |> should equal false
+        validateOp (context [("Ages", ages);])  |> should equal true
+        validateOpNotExist (context [("Ages", ages);])  |> should equal false
+    
+    [<Fact>]
+    member test.``Array comparers - invalid comparer format``() =
+        (fun () -> validator """{"Birthdays": {"$any": }}""" |> ignore) |> should throw typeof<Exception>
```

Comment:
You can keep only:
(fun () -> validator """{"Birthdays": {"$any": { "Date": { "$operatorNotExist":"value" } } }}""" |> ignore) |> should throw typeof<ParseError>
        (fun () -> validator """{"Birthdays": {"$all": { "Date": { "$operatorNotExist":"value" } } }}""" |> ignore) |> should throw typeof<ParseError>
``` (exception should be type ParseError.
The null case won't fail has it means the array has a value with null or all values are null, weird, but not an error (consistent with the matcher spec).
The other case of invalid json string is not actually testing real code as the json/string parsing is done by the 'validator' function defined in the test suite. (createEvaluator get a JsonValue)

You can add additional test at Parser (I suggest 'use rule with nested matcher')